### PR TITLE
GH-446: Resolve property placeholders in the brokerProperties attribute.

### PR DESCRIPTION
**Pull request by emetsger**
Resolves https://github.com/spring-projects/spring-kafka/issues/446

### CODE REVIEW

**Comment by artembilan** [comment]
Thank you !

Any chances to demonstrate it via tests?
Also some Docs on the matter would be nice to have, too

**Comment by emetsger** [comment]
@artembilan haha I thought you might ask for tests!  I'll work on some, along with doco!  Thanks for considering the PR!

**Comment by emetsger** [comment]
@artembilan I added a simple unit test and some documentation.

The unit test could be made more thorough, but that would require some more invasive modifications, exposing state to the test that is currently not available to it.  Or maybe you have some suggestions?

I added some documentation, too.  Let me know if it needs to be more expansive or if it's too wordy.

**Comment by emetsger** [comment]
Working on checkstyle fixes.

**Comment by emetsger** [comment]
@artembilan ready for review!

File: src/reference/asciidoc/testing.adoc [hunk 1]
```diff
@@ -150,6 +150,19 @@ public class KafkaStreamsTests {
...
+The `brokerProperties` attribute of `@EmbeddedKafka` supports standard Spring property placeholder resolution from the `Environment`.  For example:
+[source, java]
+----
+@TestPropertySource(locations = { "classpath:/kafka-broker.properties" })
+@EmbeddedKafka(topics = { "any-topic" },
+		brokerProperties = { "log.dir=${kafka.broker.logs-dir}", "listeners=PLAINTEXT://localhost:${kafka.broker.port}", "auto.create.topics.enable=true" })
+----
+In this example, the property placeholders `${kafka.broker.logs-dir}` and `${kafka.broker.port}` are resolved from the Spring `Environment`.   The `Environment` contains system properties, the system environment, _and_ properties from the classpath resource `/kafka-broker.properties`.  All of these property sources are consulted when resolving property placeholders.  If the property placeholder is resolved by the `Environment`, the resolved value will be supplied to the broker, otherwise the original value will be supplied.
+
+*_N.B._*: Property placeholders are _not_ supported for other attributes of `@EmbeddedKafka` at this time.
```
Comment (artembilan) [inline src/reference/asciidoc/testing.adoc]: We don't need to tell this.
The `EmbeddedKafka.brokerProperties()` JavaDoc has to mention that it is aware about PP resolution.

File: spring-kafka-test/src/test/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizerTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+					Assert.isTrue(expectedResults.containsKey(propertyString),
+							"Unexpected property string '" + propertyString + "'"); // sanity
+					return expectedResults.get(propertyString);
+				});
+
+		EmbeddedKafkaContextCustomizer underTest = new EmbeddedKafkaContextCustomizer(embeddedKafka);
+		underTest.customizeContext(context, mock(MergedContextConfiguration.class));
+
+		// Verify that an attempt was made to resolve each property string
+		then(env).should(times(expectedResults.size())).resolvePlaceholders(anyString());
```
Comment (artembilan) [inline spring-kafka-test/src/test/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizerTest.java]: I somehow don't understand this test - it isn't so simple as you mention.

There is `KafkaStreamsTests` where we can modify its `brokerProperties` like this:
```
brokerProperties = {
				"auto.create.topics.enable=${topics.autoCreate:false}",
				"delete.topic.enable=${topic.delete:true}" }
```
And PP resolver will take care properly about nonexistent properties and fall back to the defaults after column.

Comment (emetsger) [reply spring-kafka-test/src/test/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizerTest.java]: Hi @artembilan,

When I make the following change:
```diff
diff --git a/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java b/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java
index 6014807..4f7ee2a 100644
--- a/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java
+++ b/spring-kafka/src/test/java/org/springframework/kafka/kstream/KafkaStreamsTests.java
@@ -78,8 +78,8 @@ import org.springframework.util.concurrent.SettableListenableFuture;
                                KafkaStreamsTests.STREAMING_TOPIC2,
                                KafkaStreamsTests.FOOS },
                brokerProperties = {
-                               "auto.create.topics.enable=false",
-                               "delete.topic.enable=true" })
+                               "auto.create.topics.enable=${topics.autoCreate:false}",
+                               "delete.topic.enable=${topic.delete:true}" })
 public class KafkaStreamsTests {
 
        static final String STREAMING_TOPIC1 = "streamingTopic1";
```
`KafkaStreamsTests` passes.  If I change `${topics.autoCreate:false}` to `${topics.autoCreate:true}`, the test fails as expected (due to the assertion on line 106).  

So it seems to me that the PR works correctly in this case.

Comment (artembilan) [reply spring-kafka-test/src/test/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizerTest.java]: Exactly! 
That's what I meant.
And that's why I told you that we don't need your hacky test - we are easily covered by the simple modification to the existing test.

Comment (emetsger) [reply spring-kafka-test/src/test/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizerTest.java]: @artembilan AH I see now!  Ok, great!

Comment (emetsger) [reply spring-kafka-test/src/test/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizerTest.java]: Sorry, just misunderstanding on my part!

**Comment by emetsger** [comment]
@artembilan Ok!  So with that misunderstanding cleared up, the remaining change is to add a `brokerPropertiesLocation` attribute to `@EmbeddedKafka` correct?

**Comment by artembilan** [comment]
> add a `brokerPropertiesLocation` attribute to `@EmbeddedKafka`

Sounds good to me!

Thank you for consideration!

File: spring-kafka-test/src/main/java/org/springframework/kafka/test/context/EmbeddedKafka.java [hunk 3]
```diff
@@ -90,7 +90,8 @@
 
 	/**
 	 * Properties in form {@literal key=value} that should be added
-	 * to the broker config before runs.
+	 * to the broker config before runs.  Properties may contain property placeholders.
+	 * that will be resolved by the Spring {@code Environment}.
```
Comment (artembilan) [inline spring-kafka-test/src/main/java/org/springframework/kafka/test/context/EmbeddedKafka.java]: The sentence must start from Capital letter and should be in the present time:

> They are resolved against the Spring `{@code Environment}`
including `{@code @TestPropertySource}`.

File: spring-kafka-test/src/main/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizer.java [hunk 4]
```diff
@@ -64,7 +64,7 @@ public void customizeContext(ConfigurableApplicationContext context, MergedConte
 				continue;
 			}
 			try {
-				properties.load(new StringReader(pair));
+				properties.load(new StringReader(context.getEnvironment().resolvePlaceholders(pair)));
```
Comment (artembilan) [inline spring-kafka-test/src/main/java/org/springframework/kafka/test/context/EmbeddedKafkaContextCustomizer.java]: Nit-pick: extract `context.getEnvironment()` into the variable before the loop.

File: src/reference/asciidoc/testing.adoc [hunk 5]
```diff
@@ -150,6 +150,17 @@ public class KafkaStreamsTests {
...
+'''
+
+The `brokerProperties` attribute of `@EmbeddedKafka` supports standard Spring property placeholder resolution from the `Environment`.  For example:
+[source, java]
+----
+@TestPropertySource(locations = { "classpath:/kafka-broker.properties" })
+@EmbeddedKafka(topics = { "any-topic" },
+		brokerProperties = { "log.dir=${kafka.broker.logs-dir}", "listeners=PLAINTEXT://localhost:${kafka.broker.port}", "auto.create.topics.enable=true" })
+----
+In this example, the property placeholders `${kafka.broker.logs-dir}` and `${kafka.broker.port}` are resolved from the Spring `Environment`.   The `Environment` contains system properties, the system environment, _and_ properties from the classpath resource `/kafka-broker.properties`.  All of these property sources are consulted when resolving property placeholders.  If the property placeholder is resolved by the `Environment`, the resolved value will be supplied to the broker, otherwise the original value will be supplied.
```
Comment (artembilan) [inline src/reference/asciidoc/testing.adoc]: Would be better to start each sentence from new line: http://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line.
And I don't see reason in the last one: that is really obvious that resolved values will go to the embedded broker. That is definitely the point of this `brokerProperties` option.

**Comment by emetsger** [comment]
@artembilan ready for review!  Not sure why Travis failed, builds succeed locally:
```
esm:~/workspaces/rmap/spring-kafka (GH-446)$ ./gradlew clean test

> Task :spring-kafka:test
objc[12908]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.

> Task :spring-kafka-test:test
objc[12915]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.


BUILD SUCCESSFUL in 1m 40s
12 actionable tasks: 11 executed, 1 up-to-date
esm:~/workspaces/rmap/spring-kafka (GH-446)$ ./gradlew check --no-daemon

BUILD SUCCESSFUL in 7s
13 actionable tasks: 4 executed, 9 up-to-date
esm:~/workspaces/rmap/spring-kafka (GH-446)$ ./gradlew clean check --no-daemon

> Task :spring-kafka:test
objc[12950]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.

> Task :spring-kafka-test:test
objc[12953]: Class JavaLaunchHelper is implemented in both /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/bin/java and /Library/Java/JavaVirtualMachines/jdk1.8.0_73.jdk/Contents/Home/jre/lib/libinstrument.dylib. One of the two will be used. Which one is undefined.


BUILD SUCCESSFUL in 2m 51s
16 actionable tasks: 15 executed, 1 up-to-date
esm:~/workspaces/rmap/spring-kafka (GH-446)$ git status
On branch GH-446
Your branch is up-to-date with 'origin/GH-446'.
nothing to commit, working directory clean
esm:~/workspaces/rmap/spring-kafka (GH-446)$ git branch -aav|grep GH-446
* GH-446                                         db8474f GH-446: Support `brokerPropertiesLocation` attribute on EmbeddedKafka.
  remotes/origin/GH-446                          db8474f GH-446: Support `brokerPropertiesLocation` attribute on EmbeddedKafka.
esm:~/workspaces/rmap/spring-kafka (GH-446)$ 

```

**Comment by artembilan** [comment]
```
org.springframework.kafka.annotation.EnableKafkaIntegrationTests > testBatch FAILED
    org.mockito.exceptions.misusing.UnfinishedStubbingException at EnableKafkaIntegrationTests.java:344
```

That's that kind of nasty sporadic issues with partial Mockito stubbing we might miss at glance during development. 

I agree, that is already different story and out side of this PR, but I will be glad  if you are able to fix that as well.

Meanwhile reviewing your changes...

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-kafka/commit/0634b05c8e122f1cca70555bb551e23535549f06 with the subsequent polishing via https://github.com/spring-projects/spring-kafka/commit/e5a564e89cd6da347ca9edd0ffb97a2500c7dfc9.

@emetsger ,

Thank you again for the contribution.
Looking forward for more!
